File: V2/DIDx509/Parsing/PercentEncoding.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

namespace DIDx509.Parsing;

using System;
using System.Text;

/// <summary>
/// Provides percent-encoding and decoding according to RFC 3986 and DID:X509 spec.
/// Allowed unencoded characters: ALPHA / DIGIT / "-" / "." / "_"
/// Note: Tilde (~) is NOT allowed unencoded per DID:X509 spec (differs from RFC 3986).
/// </summary>
public static class PercentEncoding
{
    internal static class ClassStrings
    {
        internal const string ByteToUpperHex2Format = "X2";
    }

    /// <summary>
    /// Encodes a string using percent-encoding per DID:X509 spec.
    /// Only ALPHA, DIGIT, '-', '.', '_' are allowed unencoded.
    /// </summary>
    /// <param name="value">The value to encode.</param>
    /// <returns>The encoded value.</returns>
    public static string Encode(string value)
    {
        if (string.IsNullOrEmpty(value))
        {
            return string.Empty;
        }

        var encoded = new StringBuilder(value.Length * 2);

        foreach (char c in value)
        {
            if (IsDidX509AllowedCharacter(c))
            {
                encoded.Append(c);
            }
            else
            {
                // Encode as UTF-8 bytes
                byte[] bytes = Encoding.UTF8.GetBytes(new[] { c });
                foreach (byte b in bytes)
                {
                    encoded.Append(DidX509Constants.PercentChar);
                    encoded.Append(b.ToString(ClassStrings.ByteToUpperHex2Format));
                }
            }
        }

        return encoded.ToString();
    }

    /// <summary>
    /// Decodes a percent-encoded string.
    /// </summary>
    /// <param name="value">The value to decode.</param>
    /// <returns>The decoded value.</returns>
    public static string Decode(string value)
    {
        if (string.IsNullOrEmpty(value))
        {
            return string.Empty;
        }

        if (!value.Contains(DidX509Constants.PercentChar.ToString()))
        {
            return value;
        }

        var bytes = new System.Collections.Generic.List<byte>();
        var result = new StringBuilder(value.Length);

        for (int i = 0; i < value.Length; i++)
        {
            char c = value[i];

            if (c == DidX509Constants.PercentChar && i + 2 < value.Length)
            {
                // Decode percent-encoded byte
                string hex = value.Substring(i + 1, 2);
                if (IsHexDigit(hex[0]) && IsHexDigit(hex[1]))
                {
                    byte b = Convert.ToByte(hex, 16);
                    bytes.Add(b);
                    i += 2;
                    continue;
                }
            }

            // Flush accumulated bytes if any
            if (bytes.Count > 0)
            {
                result.Append(Encoding.UTF8.GetString(bytes.ToArray()));
                bytes.Clear();
            }

            // Append non-encoded character
            result.Append(c);
        }

        // Flush remaining bytes
        if (bytes.Count > 0)
        {
            result.Append(Encoding.UTF8.GetString(bytes.ToArray()));
        }

        return result.ToString();
    }

    /// <summary>
    /// Checks if a character is allowed unencoded in DID:X509.
    /// Per spec: ALPHA / DIGIT / "-" / "." / "_"
    /// </summary>
    /// <param name="c">The character to check.</param>
    /// <returns><see langword="true"/> if the character is allowed; otherwise, <see langword="false"/>.</returns>
    public static bool IsDidX509AllowedCharacter(char c)
    {
        return (c >= 'A' && c <= 'Z') ||
               (c >= 'a' && c <= 'z') ||
               (c >= '0' && c <= '9') ||
               c == DidX509Constants.HyphenChar ||
               c == DidX509Constants.UnderscoreChar ||
               c == DidX509Constants.PeriodChar;
    }

    private static bool IsHexDigit(char c)
    {
        return (c >= '0' && c <= '9') ||
               (c >= 'a' && c <= 'f') ||
               (c >= 'A' && c <= 'F');
    }
}

File: native/rust/did/x509/src/parsing/percent_encoding.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

use crate::error::DidX509Error;

/// Percent-encodes a string according to DID:x509 specification.
/// Only ALPHA, DIGIT, '-', '.', '_' are allowed unencoded.
/// Note: Tilde (~) is NOT allowed unencoded per DID:x509 spec (differs from RFC 3986).
pub fn percent_encode(input: &str) -> String {
    if input.is_empty() {
        return String::new();
    }

    let mut encoded = String::with_capacity(input.len() * 2);

    for ch in input.chars() {
        if is_did_x509_allowed_character(ch) {
            encoded.push(ch);
        } else {
            // Encode as UTF-8 bytes
            let mut buf = [0u8; 4];
            let bytes = ch.encode_utf8(&mut buf).as_bytes();
            for &byte in bytes {
                encoded.push('%');
                encoded.push_str(&format!("{:02X}", byte));
            }
        }
    }

    encoded
}

/// Percent-decodes a string.
pub fn percent_decode(input: &str) -> Result<String, DidX509Error> {
    if input.is_empty() {
        return Ok(String::new());
    }

    if !input.contains('%') {
        return Ok(input.to_string());
    }

    let mut bytes = Vec::new();
    let mut result = String::with_capacity(input.len());
    let chars: Vec<char> = input.chars().collect();
    let mut i = 0;

    while i < chars.len() {
        let ch = chars[i];

        if ch == '%' && i + 2 < chars.len() {
            let hex1 = chars[i + 1];
            let hex2 = chars[i + 2];

            if is_hex_digit(hex1) && is_hex_digit(hex2) {
                let hex_str = format!("{}{}", hex1, hex2);
                let byte = u8::from_str_radix(&hex_str, 16)
                    .map_err(|_| DidX509Error::PercentDecodingError(format!("Invalid hex: {}", hex_str)))?;
                bytes.push(byte);
                i += 3;
                continue;
            }
        }

        // Flush accumulated bytes if any
        if !bytes.is_empty() {
            let decoded = String::from_utf8(bytes.clone())
                .map_err(|e| DidX509Error::PercentDecodingError(format!("Invalid UTF-8: {}", e)))?;
            result.push_str(&decoded);
            bytes.clear();
        }

        // Append non-encoded character
        result.push(ch);
        i += 1;
    }

    // Flush remaining bytes
    if !bytes.is_empty() {
        let decoded = String::from_utf8(bytes)
            .map_err(|e| DidX509Error::PercentDecodingError(format!("Invalid UTF-8: {}", e)))?;
        result.push_str(&decoded);
    }

    Ok(result)
}

/// Checks if a character is allowed unencoded in DID:x509.
/// Per spec: ALPHA / DIGIT / "-" / "." / "_"
pub fn is_did_x509_allowed_character(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '-' || c == '.' || c == '_'
}

fn is_hex_digit(c: char) -> bool {
    c.is_ascii_hexdigit()
}
